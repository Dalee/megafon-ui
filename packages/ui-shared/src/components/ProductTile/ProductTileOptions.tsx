import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate, Tooltip } from '@megafon/ui-core';
import './style/ProductTileOptions.less';
import { IOption } from './ProductTile';

export interface IProductTileOptionsProps {
    /** Has head */
    head?: string;
    /** Options */
    options: IOption[];
    /** Handle bubble */
    onClickBubble?: () => void;
}

const cn = cnCreate('mfui-beta-product-tile-options');
class ProductTileOptions extends React.Component<IProductTileOptionsProps> {
    static propTypes = {
        head: PropTypes.string,
        options: PropTypes.arrayOf(PropTypes.shape({
            title: PropTypes.string,
            caption: PropTypes.string,
            value: PropTypes.string,
            unit: PropTypes.string,
            footnote: PropTypes.string,
            svgIcon: PropTypes.element,
        })),
        onClickBubble: PropTypes.func,
    };

    constructor(props: IProductTileOptionsProps) {
        super(props);

        const { options } = this.props;

        options && options.forEach((option, index) => {
            if (option.footnote) {
                this[`option${index}footnote`] = React.createRef<HTMLDivElement>();
            }
        });
    }

    renderFootnote(title: string, footnote: string, footnoteRef: React.RefObject<HTMLDivElement>): JSX.Element {
        const { onClickBubble } = this.props;

        return (
            <div className={cn('content')}>
                <div className={cn('title')} dangerouslySetInnerHTML={{ __html: title }} />
                <div className={cn('cashback-c')}>
                    <div
                        ref={footnoteRef}
                        className={cn('bubble-trigger')}
                        onClick={onClickBubble}
                    >
                        Подробнее
                    </div>
                    <Tooltip
                        placement="bottom"
                        triggerEvent="click"
                        triggerElement={footnoteRef}
                    >
                        <div
                            className={cn('hint-text')}
                            dangerouslySetInnerHTML={{ __html: footnote }}
                        />
                    </Tooltip>
                </div>
            </div>
        );
    }

    renderContent(title: string, caption: string = '', value: string = '', unit: string = ''): JSX.Element {
        return (
            <div className={cn('content')}>
                <div className={cn('title')} dangerouslySetInnerHTML={{ __html: title }} />
                <div className={cn('description')}>
                    {caption || `${value} ${unit}`}
                </div>
            </div>
        );
    }

    renderIcon(svgIcon: JSX.Element | null = null): JSX.Element {
        return (
            <div className={cn('icon')}>{svgIcon}</div>
        );
    }

    render() {
        const { options, head } = this.props;

        return (
            <div className={cn('')}>
                {!!head && <div className={cn('head')}>{head}</div>}
                <div className={cn('wrapper')}>
                    {options.map((option, index) => {
                        const { title, caption, value, unit, svgIcon, footnote } = option;

                        return (
                            <div className={cn('option')} key={title + index}>
                                {this.renderIcon(svgIcon)}
                                {footnote
                                    ? this.renderFootnote(title, footnote, this[`option${index}footnote`])
                                    : this.renderContent(title, caption, value, unit)
                                }
                            </div>
                        );
                    })}
                </div>
            </div>
        );
    }
}

export default ProductTileOptions;
