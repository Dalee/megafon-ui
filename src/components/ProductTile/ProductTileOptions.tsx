import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from '../../utils/cn';
import './style/ProductTileOptions.less';
import BubbleHint from '../BubbleHint/BubbleHint';
import { IOption } from './ProductTile';

interface IProductTileOptionsProps {
    /** Has head */
    head?: boolean;
    /** Options */
    options: Array<Partial<IOption>>;
    /** Handle bubble */
    onClickBubble?(): void;
}

const cn = cnCreate('mfui-product-tile-options');
class ProductTileOptions extends React.Component<IProductTileOptionsProps, {}> {
    static propTypes = {
        head: PropTypes.bool,
        options: PropTypes.arrayOf(PropTypes.shape({
            title: PropTypes.string,
            caption: PropTypes.stirng,
            value: PropTypes.number,
            unit: PropTypes.string,
            footnote: PropTypes.string,
            svgIcon: PropTypes.element,
        })),
        onClickBubble: PropTypes.func,
    };

    renderFootnote(title: string, footnote: string): JSX.Element {
        const { onClickBubble } = this.props;

        return (
            <div className={cn('title')}>
                {title}
                <BubbleHint
                    className={cn('cashback-c')}
                    popupWidth="small"
                    placement="bottom"
                    click
                    trigger={<span className={cn('bubble-trigger')} onClick={onClickBubble}>Подробнее</span>}
                >
                    <div
                        className={cn('hint-text')}
                        dangerouslySetInnerHTML={{ __html: footnote }}
                    />
                </BubbleHint>
            </div>
        );
    }

    renderContent(title: string, caption: string, value: number, unit: string): JSX.Element {
        return (
            <div className={cn('content')}>
                <div className={cn('title')}>
                    {title}
                </div>
                <div className={cn('description')}>
                    {caption ? caption : `${value} ${unit}`}
                </div>
            </div>
        );
    }

    renderIcon(svgIcon: JSX.Element): JSX.Element {
        return (
            <div className={cn('icon')}>{svgIcon}</div>
        );
    }

    render() {
        const { options, head } = this.props;

        return (
            <div className={cn('')}>
                {head && <div className={cn('head')}>Доступные опции:</div>}
                <div className={cn('wrapper')}>
                    {options.map((option, index) => {
                        const { title, caption, value, unit, svgIcon, footnote } = option;

                        return (
                            <div className={cn('option')} key={title! + index}>
                                {this.renderIcon(svgIcon!)}
                                {footnote
                                    ? this.renderFootnote(title!, footnote)
                                    : this.renderContent(title!, caption!, value!, unit!)
                                }
                            </div>
                        );
                    })}
                </div>
            </div>
        );
    }
}

export default ProductTileOptions;
