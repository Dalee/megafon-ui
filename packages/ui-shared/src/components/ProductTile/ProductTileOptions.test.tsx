import * as React from 'react';
import { shallow } from 'enzyme';
import ProductTileOptions from './ProductTileOptions';
import MegafonTv from 'icons/Service-logos/24/MF-TV.svg';

const props = {
    head: 'head',
    options: [
        {
            title: 'title',
            caption: 'caption',
            value: 'value',
            unit: 'unit',
            footnote: 'footnote',
            svgIcon: <MegafonTv />,
        },
        {
            title: 'title1',
            caption: 'caption1',
            value: 'value1',
            unit: 'unit1',
        },
        {
            title: 'title2',
            value: 'value2',
            unit: 'unit2',
        },
        {
            title: 'title3',
            value: 'value3',
        },
        {
            title: 'title4',
        },
    ],
};

describe('<ProductTileOptions />', () => {
    beforeEach(() => {
        jest.spyOn(global.Math, 'random').mockReturnValue(0.1234567);
    });
    afterEach(() => {
        jest.spyOn(global.Math, 'random').mockRestore();
    });

    it('it renders ProductTileOptions', () => {
        const wrapper = shallow(<ProductTileOptions options={[]} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders ProductTileOptions with props', () => {
        const wrapper = shallow(<ProductTileOptions {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});
