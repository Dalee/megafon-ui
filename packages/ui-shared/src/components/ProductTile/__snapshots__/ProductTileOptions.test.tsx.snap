// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTileOptions /> it renders ProductTileOptions 1`] = `
<div
  className="mfui-beta-product-tile-options"
>
  <div
    className="mfui-beta-product-tile-options__wrapper"
  />
</div>
`;

exports[`<ProductTileOptions /> it renders ProductTileOptions with props 1`] = `
<div
  className="mfui-beta-product-tile-options"
>
  <div
    className="mfui-beta-product-tile-options__head"
  >
    head
  </div>
  <div
    className="mfui-beta-product-tile-options__wrapper"
  >
    <div
      className="mfui-beta-product-tile-options__option"
      key="title0"
    >
      <div
        className="mfui-beta-product-tile-options__icon"
      >
        <test-file-stub />
      </div>
      <div
        className="mfui-beta-product-tile-options__content"
      >
        <div
          className="mfui-beta-product-tile-options__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "title",
            }
          }
        />
        <div
          className="mfui-beta-product-tile-options__footnote"
          key="297d18"
        >
          <div
            className="mfui-beta-product-tile-options__footnote-trigger"
          >
            Подробнее
          </div>
          <Tooltip
            placement="bottom"
            triggerElement={
              Object {
                "current": null,
              }
            }
            triggerEvent="click"
          >
            <div
              className="mfui-beta-product-tile-options__footnote-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "footnote",
                }
              }
            />
          </Tooltip>
        </div>
      </div>
    </div>
    <div
      className="mfui-beta-product-tile-options__option"
      key="title11"
    >
      <div
        className="mfui-beta-product-tile-options__content"
      >
        <div
          className="mfui-beta-product-tile-options__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "title1",
            }
          }
        />
        <div
          className="mfui-beta-product-tile-options__caption"
        >
          caption1
        </div>
      </div>
    </div>
    <div
      className="mfui-beta-product-tile-options__option"
      key="title22"
    >
      <div
        className="mfui-beta-product-tile-options__content"
      >
        <div
          className="mfui-beta-product-tile-options__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "title2",
            }
          }
        />
        <div
          className="mfui-beta-product-tile-options__caption"
        >
          value2 unit2
        </div>
      </div>
    </div>
    <div
      className="mfui-beta-product-tile-options__option"
      key="title33"
    >
      <div
        className="mfui-beta-product-tile-options__content"
      >
        <div
          className="mfui-beta-product-tile-options__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "title3",
            }
          }
        />
        <div
          className="mfui-beta-product-tile-options__caption"
        >
          value3 
        </div>
      </div>
    </div>
    <div
      className="mfui-beta-product-tile-options__option"
      key="title44"
    >
      <div
        className="mfui-beta-product-tile-options__content"
      >
        <div
          className="mfui-beta-product-tile-options__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "title4",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
